lib/spack/spack/cmd/ci.py

```diff
@@ -0,0 +1,478 @@
+    # entries to S3 using the boto3 api.  We import the SPACK_SIGNING_KEY using
+    # the "gpg2 --import" command, it is used both for verifying dependency
+    # buildcache entries and signing the buildcache entry we create for our
+    # target pkg.
+    #
+    # AWS_ACCESS_KEY_ID
+    # AWS_SECRET_ACCESS_KEY
+    # AWS_ENDPOINT_URL (only needed for non-AWS S3 implementations)
+    # SPACK_SIGNING_KEY
+
+    cdash_report_dir = os.path.join(ci_artifact_dir, 'cdash_report')
+    temp_dir = os.path.join(ci_artifact_dir, 'jobs_scratch_dir')
+    job_log_dir = os.path.join(temp_dir, 'logs')
+    spec_dir = os.path.join(temp_dir, 'specs')
+
+    local_mirror_dir = os.path.join(ci_artifact_dir, 'local_mirror')
+    build_cache_dir = os.path.join(local_mirror_dir, 'build_cache')
+
+    artifact_mirror_url = 'file://' + local_mirror_dir
+
+    # Clean out scratch directory from last stage
+    if os.path.exists(temp_dir):
+        shutil.rmtree(temp_dir)
+
+    if os.path.exists(cdash_report_dir):
+        shutil.rmtree(cdash_report_dir)
+
+    os.makedirs(job_log_dir)
+    os.makedirs(spec_dir)
+
+    job_spec_yaml_path = os.path.join(
+        spec_dir, '{0}.yaml'.format(job_spec_pkg_name))
+    job_log_file = os.path.join(job_log_dir, 'cdash_log.txt')
+
+    cdash_build_id = None
+    cdash_build_stamp = None
+
+    with open(job_log_file, 'w') as log_fd:
+        os.dup2(log_fd.fileno(), sys.stdout.fileno())
+        os.dup2(log_fd.fileno(), sys.stderr.fileno())
+
+        tty.msg('job concrete spec path: {0}'.format(job_spec_yaml_path))
+
+        spack_ci.import_signing_key(signing_key)
+
+        real_compilers = spack_ci.configure_compilers(compiler_action)
+
+        spec_map = spack_ci.get_concrete_specs(
+            root_spec, job_spec_pkg_name, related_builds, compiler_action,
+            real_compilers)
+
+        job_spec = spec_map[job_spec_pkg_name]
+
+        tty.msg('Here is the concrete spec: {0}'.format(job_spec))
+
+        with open(job_spec_yaml_path, 'w') as fd:
+            fd.write(job_spec.to_yaml(hash=ht.build_hash))
+
+        tty.msg('Done writing concrete spec')
+
+        # DEBUG
+        with open(job_spec_yaml_path) as fd:
+            tty.msg('Just wrote this file, reading it, here are the contents:')
+            tty.msg(fd.read())
+
+        # DEBUG the root spec
+        root_spec_yaml_path = os.path.join(spec_dir, 'root.yaml')
+        with open(root_spec_yaml_path, 'w') as fd:
+            fd.write(spec_map['root'].to_yaml(hash=ht.build_hash))
+
+        if bindist.needs_rebuild(job_spec, remote_mirror_url, True):
+            # Binary on remote mirror is not up to date, we need to rebuild
+            # it.
+            #
+            # FIXME: ensure mirror precedence causes this local mirror to
+            # be chosen ahead of the remote one when installing deps
+            if enable_artifacts_mirror:
+                spack_cmd('mirror', 'add', 'local_mirror', artifact_mirror_url)
```

Comment:
I hope you don't mind me including a request here and I understand if you don't think it would fit in the effort.
For containers adding a mirror makes sense but if we use the pipeline in the context of a runner deployed at a site this would change the user's `~/.spack/mirrors.yaml` file. As part of a CI job this may not exists and if I'm running multiple CI jobs concurrently the value could be overwritten and lead to unexpected job failures.

There may be a better way but what about defining all the job specific configuration and supplying it in a similar method as to `--config-scope`?

Reply:
Thanks @paulbry I definitely don't mind you including a request here :smile: 

So is your question just about the use of gitlab's artifacts directory as a mirror?  If so, that is now optional and off by default (I plan to support turning it on via the `spack.yaml`, but atm it is configured with a CI env var).  Let me know if that eases your concern in this particular case.

But I think I see your point in general: in the presence of parallel CI jobs running against the same spack, we should be careful about modifying configs.  Is that a reasonable summary?

Reply:
@scottwittenburg Thanks that makes sense. Being part of the CI env would work even in the case of concurrently running jobs.